STITCH-1955 Implement NamespaceChangeStreamListener

jsflax opened the pull request:
*Not ready for review*

edaniels commented on LICENSE:
```diff
@@ -143,7 +143,7 @@ permissions under this License.
 8. Limitation of Liability.
 
-In no event and under no legal theory, whether in tort (including negligence),
+In no coreEvent and under no legal theory, whether in tort (including negligence),
```
undo this

edaniels commented on server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -699,114 +735,158 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
-            coll.sync(doc1Id, DefaultSyncConflictResolvers.localWins())
+
+            coll.configure(DefaultSyncConflictResolvers.localWins(), null, null)
+            coll.syncOne(doc1Id)
```
The point of these types of tests are that once you sync once, you don't need to request it again after this point

edaniels commented on server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -699,114 +735,158 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
 
-            coll.sync(doc1Id, DefaultSyncConflictResolvers.localWins())
-            listenAndSync() // syncs this time
+            coll.configure(DefaultSyncConflictResolvers.localWins(), null, null)
```
Based on the diff here, the comment should be that it syncs this time right? The second syncone and syncpass shouldn't be necessary

edaniels commented on server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -699,114 +735,158 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
 
             powerCycleDevice()
-            listenAndSync() // does nothing with no conflict handler
+            coll.configure(DefaultSyncConflictResolvers.localWins(), null, null)
```
there should be no configuration here, the old test doesn't configure anything.

edaniels commented on server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -699,114 +735,158 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
-            coll.sync(coll.synchronizedDocuments.toTypedArray()[0].documentId, DefaultSyncConflictResolvers.localWins())
-            listenAndSync() // resolves the conflict
+            assertEquals(1, coll.syncedIds.size)
+            coll.syncOne(coll.syncedIds.toTypedArray()[0])
```
replace this with a configure

edaniels commented on server/services/mongodb-remote/src/test/java/com/mongodb/stitch/server/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -699,114 +735,158 @@ class SyncMongoClientIntTests : BaseStitchServerIntTest() {
-            listenAndSync()
+
+            coll.configure(DefaultSyncConflictResolvers.localWins(), null, null)
+            coll.syncOne(doc1Id)
```
don't need any of these sync ones

jsflax commented:
@edaniels @adamchel PTAL

adamchel commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceChangeStreamListener.java:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  /**
+   * Opens the next next.
```
[nit] next next?

adamchel commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceChangeStreamListener.java:
```diff
@@ -0,0 +1,245 @@
+  void stream() {
+    logger.info("stream START");
+    if (!networkMonitor.isConnected()) {
+      logger.info("stream END - Network disconnected");
```
[q] does this need to trigger onNetworkStateChanged or is that already being handled?

edaniels commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+              changeEventForLocalReplace(
+                  nsConfig.getNamespace(),
+                  docId,
+                  staleDocumentMap.get(docId),
```
This should come from the result of a remote query to get the latest docs. So there should be something that indicates what doc ids are stale and then something get the latest for those.

jsflax replied:
👍

edaniels commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+        }
+
+        if (nsConfig.isStale() && staleIds.contains(docId)) {
+          if (docConfig.hasCommittedVersion(DocumentVersionInfo
```
This check should be covered in the sync remote to local.

jsflax replied:
👍

jsflax commented on android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -187,564 +204,668 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
+                Document("well", "shoot")
+            }, null, null)
+            coll.syncOne(doc1Id)
+            watchAndSync { }
```
I think that's correct. While fooling around with the locking initially, I had several methods for watching and syncing. The only two we need are `watchFor` and `syncPass`. I'll refactor appropriately.

jsflax replied:
👍

jsflax commented on android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -187,564 +204,668 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
+        }
+    }
+
+    private fun queueDisposableWatcher(watcher: Callback<ChangeEvent<BsonDocument>, Any>) {
```
We only need one helper at this point, alongside `syncPass`. I do think it should wait for a specific op though, not n events. Can be more helpful at times.

jsflax replied:
And good find on `Condition`, I haven't heard of that api before.

edaniels replied:
That's fine

jsflax replied:
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceChangeStreamListener.java:
```diff
@@ -0,0 +1,245 @@
+
+  void clearWatchers() {
+    for (int i = 0; i < callbackQueue.size(); i++) {
+      callbackQueue.remove().onComplete(
```
The failed result can allow for certain test failures that might otherwise be stalled.

edaniels replied:
Got it

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/StaleDocumentFetcher.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.bson.BsonValue;
+
+public interface StaleDocumentFetcher {
```
While I agree with the other two, this allows us to pass the DataSynchronizer around when it only needs to be used to fetch stale documents. Seems handier than passing around the entire class.

edaniels replied:
Based on other comments though, there should be no passing of data sync to other places. That's more of a code smell to do that type of thing. Only data sync needs to go get latest docs. ns config will be the one to say what is and isn't stale.

jsflax replied:
Good point

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventType.java:
```diff
@@ -0,0 +1,23 @@
+public enum EventType {
+  MESSAGE,
+  ERROR,
+  EOF
```
There's definitely some conflation of terms here, as an error would be "type" in a stream, whilst a Message would arguably be a default state. I'm amenable to changing the name of this though.

jsflax replied:
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/ServerSideEventError.java:
```diff
@@ -0,0 +1,23 @@
+
+package com.mongodb.stitch.core.internal.net;
+
+public class ServerSideEventError extends Exception {
```
Good catch

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,105 @@
+  protected final CoreEvent readEvent() throws ServerSideEventError, IOException {
+    final CoreEvent coreEvent = new CoreEvent();
+
+    while (!Thread.interrupted() && this.isActive()) {
```
It can help close the stream a bit cleaner, but it doesn't really matter, since the errors are swallowed once the stream is closed.

jsflax replied:
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -1389,7 +1492,7 @@ private void addAndStartListeningToNamespace(final MongoNamespace namespace) {
    * @param event      the change event.
    */
-  private void emitEvent(final BsonValue documentId, final ChangeEvent<BsonDocument> event) {
+  public void emitEvent(final BsonValue documentId, final ChangeEvent<BsonDocument> event) {
```
It did for the `EventEmitter` interface.

jsflax commented on android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -79,34 +86,43 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
     }
 
-    private fun getTestColl(): SyncMongoCollection<Document> {
+    private fun getTestColl(): Sync<Document> {
```
👍

jsflax commented on android/services/mongodb-remote/src/androidTest/java/com/mongodb/stitch/android/services/mongodb/remote/SyncMongoClientIntTests.kt:
```diff
@@ -118,32 +134,36 @@ class SyncMongoClientIntTests : BaseStitchAndroidIntTest() {
     @Test
-    fun testWatch() {
-        testSyncInBothDirections({
+    fun testStale() {
```
👍

jsflax commented on android/services/mongodb-remote/src/main/java/com/mongodb/stitch/android/services/mongodb/remote/internal/RemoteMongoClientImpl.java:
```diff
@@ -43,4 +45,8 @@ public RemoteMongoClientImpl(
     return new RemoteMongoDatabaseImpl(proxy.getDatabase(databaseName), dispatcher);
   }
+
+  public DataSynchronizer getDataSynchronizer() {
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/EventEmitter.java:
```diff
@@ -0,0 +1,24 @@
+import org.bson.BsonValue;
+
+public interface EventEmitter {
+  void emitEvent(final BsonValue documentId, final ChangeEvent<BsonDocument> event);
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java:
```diff
@@ -58,8 +59,10 @@
 
   private NamespaceListenerConfig namespaceListenerConfig;
-  private ConflictHandler conflictHandler;
+  private AtomicReference<ConflictHandler> conflictHandler = new AtomicReference<>();
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java:
```diff
@@ -191,7 +195,32 @@ public CoreDocumentSynchronizationConfig getSynchronizedDocument(final BsonValue
   }
 
-  public CoreDocumentSynchronizationConfig addSynchronizedDocument(
+  void setStale() {
```
Sounds good, tackling this now.

jsflax replied:
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -1026,10 +1126,11 @@ private static Object resolveConflictWithResolver(
    */
-  public void doListenerSweep() {
-    instanceChangeStreamListener.sweep();
+  public void queueDisposableWatcher(final Callback<ChangeEvent<BsonDocument>, Object> watcher) {
```
This is only used in testing

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -148,19 +152,58 @@ public DataSynchronizer(
     }
     // TODO: Add back in
-    // this.instanceChangeStreamListener.start();
+    this.instanceChangeStreamListener.start();
```
I don't think that makes sense for this PR, since that's another ticket.

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceSynchronizationConfig.java:
```diff
@@ -58,8 +59,10 @@
-  private ConflictHandler conflictHandler;
+  private AtomicReference<ConflictHandler> conflictHandler = new AtomicReference<>();
   private Codec documentCodec;
+  private Set<BsonValue> staleIds;
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/common/Stream.java:
```diff
@@ -0,0 +1,60 @@
+import org.bson.codecs.configuration.CodecRegistry;
+
+
+public class Stream<T> {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/CoreEvent.java:
```diff
@@ -0,0 +1,47 @@
+
+package com.mongodb.stitch.core.internal.net;
+
+public class CoreEvent {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/Event.java:
```diff
@@ -0,0 +1,69 @@
+import org.bson.Document;
+import org.bson.codecs.Decoder;
+
+public final class Event<T> {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/common/Stream.java:
```diff
@@ -0,0 +1,60 @@
+    this.decoder = codecRegistry.get(resultClass);
+  }
+
+  public T next() throws Exception {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,105 @@
+  protected final CoreEvent readEvent() throws ServerSideEventError, IOException {
+    final CoreEvent coreEvent = new CoreEvent();
+
+    while (!Thread.interrupted() && this.isActive()) {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,105 @@
+      }
+    }
+
+    coreEvent.setType(EventType.EOF);
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,105 @@
+    return coreEvent;
+  }
+
+  private EventType doReadEventType() throws ServerSideEventError, IOException {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,105 @@
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+public abstract class EventStreamReader {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/OkHttpTransport.java:
```diff
@@ -64,7 +66,7 @@ public OkHttpTransport() {
   }
 
-  private static Response handleResponse(final okhttp3.Response response) {
+  static Response handleResponse(final okhttp3.Response response) {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/OkHttpTransport.java:
```diff
@@ -90,4 +92,23 @@ public Response roundTrip(final Request request) throws IOException {
+
+  @Override
+  public EventStream stream(final Request request) throws IOException {
+    request.getHeaders().put(
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/OkHttpTransport.java:
```diff
@@ -90,4 +92,23 @@ public Response roundTrip(final Request request) throws IOException {
+    request.getHeaders().put(
+        com.mongodb.stitch.core.internal.net.Headers.CONTENT_TYPE,
+        "text/event-next");
+    request.getHeaders().put("Accept", "text/event-stream");
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/ChangeEvent.java:
```diff
@@ -233,7 +232,7 @@ static BsonDocument toBsonDocument(final ChangeEvent<BsonDocument> value) {
             nsDoc.getString(ChangeEventCoder.Fields.NS_DB_FIELD).getValue(),
             nsDoc.getString(ChangeEventCoder.Fields.NS_COLL_FIELD).getValue()),
-        document.getDocument(ChangeEventCoder.Fields.DOCUMENT_KEY_FIELD),
+        document.getDocument(ChangeEventCoder.Fields.DOCUMENT_KEY_FIELD, null),
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/ChangeEvent.java:
```diff
@@ -190,7 +190,6 @@ static BsonDocument toBsonDocument(final ChangeEvent<BsonDocument> value) {
     keyPresent(ChangeEventCoder.Fields.ID_FIELD, document);
     keyPresent(ChangeEventCoder.Fields.OPERATION_TYPE_FIELD, document);
     keyPresent(ChangeEventCoder.Fields.NS_FIELD, document);
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceChangeStreamListener.java:
```diff
@@ -0,0 +1,245 @@
+      while (currentStream.isOpen()) {
+        final Event<ChangeEvent<BsonDocument>> event = currentStream.nextEvent();
+        if (event.getEventType() == EventType.ERROR) {
+          throw event.getError();
```
Maybe we should just emit this error through the DataSynchronizer?

edaniels replied:
Because of the importance of these errors. I would have maybe something like the error emitter passed into there (may need that ifc back) but you want to make sure that we make that a special emitter that can look at what went wrong. If it's an invalid session or something like that via handling the rich error, that will need to bubble up to the right place. I would make a TODO for this and file a ticket so that we can move on with this PR. Create the ticket and then do `TODO(STITCH-XXX)` on this line replacing the ticket number.

jsflax commented on core/sdk/src/test/java/com/mongodb/stitch/core/internal/net/TransportIntTests.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+}
+
+class TransportIntTests {
```
👍

jsflax commented on core/sdk/src/test/java/com/mongodb/stitch/core/internal/net/TransportIntTests.kt:
```diff
@@ -0,0 +1,95 @@
+    init {
+        Thread {
+            lock.lock()
+            val server = HttpServer.create(InetSocketAddress(8000), 0)
```
👍

jsflax commented on core/sdk/src/test/java/com/mongodb/stitch/core/internal/net/TransportIntTests.kt:
```diff
@@ -0,0 +1,95 @@
+
+        val resp = transport.roundTrip(
+                Request.Builder().withUrl(
+                        "http://localhost:8000/roundTrip"
```
👍

jsflax commented on core/sdk/src/test/java/com/mongodb/stitch/core/internal/net/TransportIntTests.kt:
```diff
@@ -0,0 +1,95 @@
+    val lock = ReentrantLock()
+
+    init {
+        Thread {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/Event.java:
```diff
@@ -0,0 +1,69 @@
+        this.error = null;
+        break;
+      case ERROR:
+        final Document error = BsonUtils.parseValue(data, Document.class);
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,105 @@
+import java.io.IOException;
+
+public abstract class EventStreamReader {
+  private static final String DATA = "ata: ";
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/OkHttpEventStream.java:
```diff
@@ -0,0 +1,59 @@
+
+import okio.BufferedSource;
+
+public class OkHttpEventStream extends EventStreamReader implements EventStream {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/OkHttpTransport.java:
```diff
@@ -90,4 +92,23 @@ public Response roundTrip(final Request request) throws IOException {
   }
+
+  @Override
+  public EventStream stream(final Request request) throws IOException {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/services/internal/CoreStitchServiceClient.java:
```diff
@@ -70,6 +72,19 @@ void callFunction(
       final Class<T> resultClass,
       final CodecRegistry codecRegistry);
 
+  <T> Stream<T> streamFunction(String name,
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/services/internal/CoreStitchServiceClientImpl.java:
```diff
@@ -54,6 +58,22 @@ public CoreStitchServiceClientImpl(
+      final String name,
+      final List<?> args) {
+    final Document body = new Document();
+    body.put("name", name);
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/services/internal/CoreStitchServiceClientImpl.java:
```diff
@@ -145,6 +165,32 @@ public void callFunction(
   }
 
+  @Override
+  public <T> Stream<T> streamFunction(final String name,
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/Event.java:
```diff
@@ -0,0 +1,69 @@
+  private Event(final EventType eventType, final String data, final Decoder<T> decoder) {
+    this.eventType = eventType;
+
+    switch (eventType) {
```
Can we post the reasoning here for URL decoding? I haven't seen any reason to URL decode the data.

jsflax commented on core/sdk/src/test/java/com/mongodb/stitch/core/internal/net/TransportIntTests.kt:
```diff
@@ -0,0 +1,95 @@
+        var x = 0
+        while (stream.isOpen) {
+            val event = stream.nextEvent()
+            println(event)
```
👍

jsflax commented on core/sdk/src/test/java/com/mongodb/stitch/core/internal/net/TransportIntTests.kt:
```diff
@@ -0,0 +1,95 @@
+import java.net.InetSocketAddress
+import java.util.concurrent.locks.ReentrantLock
+
+internal class RoundTripHandler : HttpHandler {
```
👍

jsflax commented on core/sdk/src/main/java/com/mongodb/stitch/core/services/internal/CoreStitchServiceClientImpl.java:
```diff
@@ -19,13 +19,17 @@
 import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.internal.Base64;
 
 public class CoreStitchServiceClientImpl implements CoreStitchServiceClient {
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -148,19 +152,58 @@ public DataSynchronizer(
+    });
+  }
+
+  private Set<BsonDocument> getStaleDocuments(final NamespaceSynchronizationConfig nsConfig) {
```
👍

jsflax replied:
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -326,226 +372,6 @@ public boolean doSyncPass() {
     return true;
   }
 
-  /**
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/InstanceChangeStreamListenerImpl.java:
```diff
@@ -18,80 +18,82 @@
-        poller.start();
+      for (final Map.Entry<MongoNamespace, NamespaceChangeStreamListener> streamerEntry :
+          nsStreamers.entrySet()) {
+        this.instanceConfig.getNamespaceConfig(streamerEntry.getKey()).setStaleDocumentIds(
```
👍 Given some of your other comments, we my just be able to remove this line of code entirely.

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/InstanceChangeStreamListenerImpl.java:
```diff
@@ -18,80 +18,82 @@
-    } finally {
-      instanceLock.writeLock().unlock();
+  @Override
+  public void queueDisposableWatcher(final Callback<ChangeEvent<BsonDocument>, Object> watcher) {
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -148,19 +152,58 @@ public DataSynchronizer(
+      @Override
+      public void onNetworkStateChanged() {
+        if (!networkMonitor.isConnected()) {
+          for (final MongoNamespace mongoNamespace : syncConfig.getSynchronizedNamespaces()) {
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -148,19 +152,58 @@ public DataSynchronizer(
+    return ids;
+  }
+
+  public Set<BsonValue> getStaleDocumentIds(final NamespaceSynchronizationConfig nsConfig) {
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -174,16 +217,19 @@ public void reloadConfig() {
           networkMonitor,
-          authMonitor);
+          authMonitor,
+          this
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+      final Set<BsonDocument> staleDocuments = getStaleDocuments(nsConfig);
+      final Set<BsonValue> staleIds = getStaleDocumentIds(staleDocuments);
+      final Map<BsonValue, BsonDocument> staleDocumentMap = new HashMap<>();
+      for (final BsonDocument staleDocument : staleDocuments) {
```
isn't this supposed to be latest documents now? stale check would be later on

edaniels replied:
Looks fine now.

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+        }
+      }
+
+      nsConfig.setFresh();
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+  /**
+   * Synchronizes the remote state of every requested document to be synchronized with the local
+   * state of said documents. Utilizes change streams to get "recent" updates to documents of
+   * interest. Documents that are being synchronized from the first time will be fetched via a
```
👍

jsflax commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -843,6 +669,274 @@ private void syncLocalToRemote() {
+          continue;
+        }
+
+        if (nsConfig.isStale() && staleIds.contains(docId)) {
```
👍

jsflax commented:
@edaniels PTAL. All that is left is for you to answer my question about error emission, and for me to write a test for auth failures related to opening streams. otherwise, i’ve addressed your remarks and tests are passing.

edaniels commented:
@jsflax haven't taken 2nd pass yet but I can't find your question about errors.

jsflax commented:
Basically how and where to surface errors that occur while steaming.

On Thu 4 Oct 2018, 23:11 Eric Daniels, <notifications@github.com> wrote:

> @jsflax <https://github.com/jsflax> haven't taken 2nd pass yet but I
> can't find your question about errors.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mongodb/stitch-android-sdk/pull/65#issuecomment-427186340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEJ7QKk5zptNRErx8wlzwLM6Sxbi4uV1ks5uhoelgaJpZM4W9HPz>
> .
>

edaniels commented:
Got it, found it.

edaniels commented on core/sdk/src/main/java/com/mongodb/stitch/core/auth/internal/CoreStitchAuth.java:
```diff
@@ -196,6 +197,27 @@ public Response doAuthenticatedRequest(final StitchAuthRequest stitchReq) {
+          decoder
+      );
+    } catch (final StitchServiceException ex) {
+      return handleAuthFailure(ex, stitchReq, decoder, tryRefresh);
```
This isn't actually handling errors still. It needs to be similar to handleRequestError meaning you may need to inspect the event for an error.

edaniels replied:
Try to reuse the code that exists for this case.

edaniels replied:
👍

edaniels commented on core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/EventStreamReader.java:
```diff
@@ -0,0 +1,111 @@
+   * @return the fully processed event
+   * @throws IOException if a stream is in the wrong state, IO errors can be thrown
+   */
+  protected final Event processEvent() throws IOException {
```
nice

edaniels commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/CoreDocumentSynchronizationConfig.java:
```diff
@@ -20,7 +20,8 @@
 import com.mongodb.MongoNamespace;
 import com.mongodb.client.MongoCollection;
-import com.mongodb.stitch.core.services.mongodb.remote.sync.ConflictHandler;
+import com.mongodb.lang.NonNull;
```
Use the java non null

edaniels commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -148,21 +152,43 @@ public DataSynchronizer(
         Loggers.getLogger(String.format("DataSynchronizer-%s", instanceKey));
   }
 
+  private Set<BsonDocument> getLatestDocumentFromRemote(
```
`getLatestDocumentsFromRemote` <- plural

edaniels commented on core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java:
```diff
@@ -341,209 +370,88 @@ private void syncRemoteToLocal() {
-   * @param remoteChangeEvent the remote change event to synchronize into the local database.
-   * @param localColl         the local collection where the document lives.
-   */
-  private void syncRemoteChangeEventToLocal(
```
Reminder to move this to where it was to make the diff easier. I can read it now but it will be better for future reference.
